### PR TITLE
[Feature]: Script sharing

### PR DESCRIPTION
**Pull request by melonamin**
## Description

Adds the ability to share the script as an encoded URL

## How to Test

1. On any script, invoke the contextual menu and choose "Share script."  
2. Open the link in your clipboard.  
3. Verify that the previously shared script was imported with the expected name and content.  
4. Edge cases:  
    - Malformed URL  
    - A script with this name already exists.

## Checklist

- [ ] I have added at least one test for the new feature or fixed bug, or this PR does not include any app code changes.
- [ ] I have tested the new version using the auto-generated preview URL.

<details>
  <summary>How to Find the Preview URL</summary>

  The app will be deployed to a preview URL automatically every time you push a commit to this PR.

  You can find the preview link in the "Deployments" section at the bottom of this PR:
  - Look for the section with the 🚀 rocket icon that says "This branch was successfully deployed."
  - Alternatively, look for "github-actions bot deployed to preview" in the timeline.
  - Click "View deployment" to open the preview URL.
</details>

### CODE REVIEW

File: src/features/script-explorer/script-explorer.tsx [hunk 1]
```diff
@@ -68,13 +70,13 @@ export const ScriptExplorer = memo(() => {
   /**
    * Global state
    */
-  const sqlScripts = useSqlScriptNameMap();
+  const sqlScripts = useSqlScripts();
```
Comment (mishamsk) [inline src/features/script-explorer/script-explorer.tsx:73]: this is going to be bad for performance. the special getter that only fetches names is here for a reason. With this change the whole tree will re-render on each keystroke in any script tab.

see below on how to avoid this

File: src/features/script-explorer/script-explorer.tsx [hunk 2]
```diff
@@ -85,15 +87,33 @@ export const ScriptExplorer = memo(() => {
             copyToClipboard(sqlScript.label, { showNotification: true });
           },
         },
+        {
+          label: 'Share script',
+          onClick: (sqlScript) => {
+            const scriptId = sqlScript.value;
+            const script = sqlScripts.get(scriptId);
```
Comment (mishamsk) [inline src/features/script-explorer/script-explorer.tsx:94]: you do not need a reactive state var to get to script content. You can do `useAppStore.getState().sqlScripts` to access the map at any time. You do not need reactive, you just need data here

File: src/store/app-store.tsx [hunk 3]
```diff
@@ -308,18 +308,8 @@ export function useAttachedDBLocalEntriesMap(): Map<LocalEntryId, LocalFile> {
...
-        new Map(
-          Array.from(state.sqlScripts).map(([id, script]): [SQLScriptId, string] => [
-            id,
-            script.name,
-          ]),
-        ),
-    ),
-  );
+export function useSqlScripts(): Map<SQLScriptId, SQLScript> {
+  return useAppStore(useShallow((state) => state.sqlScripts));
```
Comment (mishamsk) [inline src/store/app-store.tsx:312]: just FYI this is really meaningless, especially the `useShallow` part. to access the map you could have used `useAppStore.use.sqlScripts` which is a regular reactive access to the entire map. `useShallow` helps when you create new objects on the fly, which would have been checked via `Object.is` by React and won't "match" (would cause a render) unless `useShallow` is used. But for the map itself - it is impossible for our map to change in way that it has the same keys & same object as values...

anyways, this should be reverted back - see comment where this was used

File: src/utils/script-import-utils.ts [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+  title?: string;
+  sharedScript?: SharedScript;
+}
+
+/**
+ * Validates an encoded script without importing it
+ * @param encodedScript The encoded script from the URL
+ * @returns Validation result with the decoded script if valid
+ */
+export function validateEncodedScript(encodedScript: string): ScriptValidationResult {
```
Comment (mishamsk) [inline src/utils/script-import-utils.ts:30]: rgd the "edge case" (or not so edge) where a user already has a same named script. Our app state allows non-reactive access from anywhere, including from this helper (see comment rgd it's usage in component). So, if you up for it, we can add some logic / options for the user to choose from if there is a collision. I think at the very least we can use our `findUniquerName` helper and create a new unique name even without asking.

**Review by mishamsk — CHANGES_REQUESTED** [review]
thanks a lot, this is massive. But it will degrade performance, so I made suggestions.

File: src/utils/script-import-utils.ts (hunk 4 above)
Comment (melonamin) [reply src/utils/script-import-utils.ts:30]: Not sure I'm following. `validateEncodedScript` is used to check the payload, then I use existing `createSQLScript` which already implements `findUniqueName`

**Comment by koalakate** [comment]
import shared script modal window
image
image

File: src/utils/script-import-utils.ts (hunk 4 above)
Comment (mishamsk) [reply src/utils/script-import-utils.ts:30]: yeah, sorry, my bad, I've missed that

File: tests/integration/fixtures/utils/explorer-tree.ts [hunk 5]
```diff
@@ -265,3 +265,19 @@ export const clickExplorerTreeNodeMenuItemByName = async (
...
+ */
+export async function checkIfExplorerItemExists(
+  page: Page,
+  explorerTestIdPrefix: string,
+  name: string,
+): Promise<boolean> {
+  const allItems = getAllExplorerTreeNodes(page, explorerTestIdPrefix);
+  const itemTexts = await allItems.evaluateAll((nodes) =>
+    nodes.map((node) => node.textContent?.trim() || ''),
+  );
```
Comment (mishamsk) [inline tests/integration/fixtures/utils/explorer-tree.ts:280]: this will not work correctly all the time unfortunately. I stumbled on that myself. When context menu is open, the text content includes names of all context menu buttons... That's why `getExplorerTreeNodeByName` uses a filter that locates nested `p` tab and checks it's text, rather than the entire container one... To get the names we'll either need to add separate testid to the `p` tag directly and use it here, or make a similar locator from `allItems` that reaches into the nested `p` and then call evaluate like you did

File: tests/integration/script/script-sharing.spec.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  expect(await checkIfScriptExists('query.sql')).toBeFalsy();
+
+  await openImportSharedScriptModalViaSpotlight();
+
+  const input = page.getByTestId('import-script-url-input');
+
+  await input.fill(clipboardContent);
+  await page.getByTestId('import-script-url-submit-button').click();
+
+  await expect(scriptEditorContent).toContainText('SELECT * FROM test');
```
Comment (mishamsk) [inline tests/integration/script/script-sharing.spec.ts:39]: let's add a test that importing a script with the same name doesn't overwrite existing script. So you can continue this test function or create a second one, but basically we need to change the text of the `query.sql`, then import from url again - check that `query.sql` is the same and that a new `query_1.sql` was added with the expected content.

**Review by mishamsk — APPROVED** [review]
LGTM. See two small but actionable comments rgd tests. Other than that this is ready to go.
